aten/src/ATen/native/SummaryOps.cpp

```diff
@@ -65,4 +134,112 @@ _bincount_cpu(const Tensor& self, const Tensor& weights, int64_t minlength) {
   });
 }
 
+
+std::tuple<Tensor,Tensor> _histogram_cpu(
+    const Tensor& self,
+    int64_t bins,
+    c10::optional<ArrayRef<double>> range,
+    const Tensor& weights,
+    bool density) {
+
+  // Weights having a different shape from input is not supported yet. TO DO:
+  // Add support for weights broadcastable to input
+  bool has_weights = weights.defined();
+  Tensor flattened_weights;
+  if (has_weights) {
+    TORCH_CHECK(
+        weights.sizes() == self.sizes(),
+        "histogram only supports input and weights of the same shape");
+    flattened_weights = weights.flatten(0).contiguous();
+  }
+
+
+  // If range is not defined, compute min and max from the values in the Tensor
+  Scalar min;
+  Scalar max;
+  if (range.has_value()) {
+    min = range.value()[0];
+    max = range.value()[1];
+  } else {
+    min = self.min().item();
+    max = self.max().item();
+  }
+
+  Tensor hist = AT_DISPATCH_ALL_TYPES(
```

Comment:
Dispatching on all types and only weights being float or double seems fine, but doing it like this seems odd because the type of weight tensor is ignored.

What about validating (and documenting) that weights must be a float or double type, and then using the type of weights to determine the second template arg?

Reply:
This is done to be consistent with the already existing bincount and the custom binning implementation that uses it - having the uniform binning implementation correctly add up integral weights and the custom binning implementation cast them to double would be counterintuitive for the user.